velero-plugins/imagestream/restore.go

```diff
@@ -27,67 +27,68 @@ func (p *RestorePlugin) AppliesTo() (velero.ResourceSelector, error) {
-			if strings.HasPrefix(dockerImageReference, backupInternalRegistry) {
-				destTag := ""
-				if copyToTag {
-					destTag = ":" + tag.Tag
-				}
-				srcPath := fmt.Sprintf("docker://%s/%s/%s@%s", migrationRegistry, imageStreamUnmodified.Namespace, imageStreamUnmodified.Name, tag.Items[i].Image)
-				destPath := fmt.Sprintf("docker://%s/%s/%s%s", internalRegistry, imageStreamUnmodified.Namespace, imageStreamUnmodified.Name, destTag)
+		backupInternalRegistry := annotations[common.BackupRegistryHostname]
+		if len(internalRegistry) == 0 {
+			return nil, errors.New("backup cluster registry not found for annotation \"openshift.io/backup-registry-hostname\"")
+		}
+		migrationRegistry := input.Restore.Annotations[common.MigrationRegistry]
+		if len(migrationRegistry) == 0 {
+			return nil, errors.New("migration registry not found for annotation \"openshift.io/migration\"")
+		}
+		p.Log.Info(fmt.Sprintf("[is-restore] backup internal registry: %#v", backupInternalRegistry))
+		p.Log.Info(fmt.Sprintf("[is-restore] restore internal registry: %#v", internalRegistry))
 
-				p.Log.Info(fmt.Sprintf("[is-restore] copying from: %s", srcPath))
-				p.Log.Info(fmt.Sprintf("[is-restore] copying to: %s", destPath))
-				manifest, err := copyImageRestore(srcPath, destPath)
-				if err != nil {
-					p.Log.Info(fmt.Sprintf("[is-restore] Error copying image: %v", err))
-					return nil, err
+		for _, tag := range imageStreamUnmodified.Status.Tags {
+			p.Log.Info(fmt.Sprintf("[is-restore] Restoring tag: %#v", tag.Tag))
+			specTag := findSpecTag(imageStreamUnmodified.Spec.Tags, tag.Tag)
+			copyToTag := true
+			if specTag != nil && specTag.From != nil {
+				p.Log.Info(fmt.Sprintf("[is-restore] image tagged: %s, %s", specTag.From.Kind, specTag.From.Name))
+				// we have a tag.
+				copyToTag = false
+			}
+			// Iterate over items in reverse order so most recently tagged is copied last
+			for i := len(tag.Items) - 1; i >= 0; i-- {
+				dockerImageReference := tag.Items[i].DockerImageReference
+				if strings.HasPrefix(dockerImageReference, backupInternalRegistry) {
+					destTag := ""
+					if copyToTag {
+						destTag = ":" + tag.Tag
+					}
+					srcPath := fmt.Sprintf("docker://%s/%s/%s@%s", migrationRegistry, imageStreamUnmodified.Namespace, imageStreamUnmodified.Name, tag.Items[i].Image)
+					destPath := fmt.Sprintf("docker://%s/%s/%s%s", internalRegistry, imageStreamUnmodified.Namespace, imageStreamUnmodified.Name, destTag)
+
+					p.Log.Info(fmt.Sprintf("[is-restore] copying from: %s", srcPath))
+					p.Log.Info(fmt.Sprintf("[is-restore] copying to: %s", destPath))
+					manifest, err := copyImageRestore(srcPath, destPath)
+					if err != nil {
+						p.Log.Info(fmt.Sprintf("[is-restore] Error copying image: %v", err))
+						return nil, err
+					}
+					p.Log.Info(fmt.Sprintf("[is-restore] manifest of copied image: %s", manifest))
 				}
-				p.Log.Info(fmt.Sprintf("[is-restore] manifest of copied image: %s", manifest))
 			}
 		}
 	}
-
```

Comment:
In this case, When `input.Restore.Annotations[common.MigrateTypeAnnotation] == ""`, we may not want to use the same return behavior, since it returns with the flag set by `WithoutRestore()` -- when we're managing the image copy, we don't actually restore the ImageStream resource, since for local images the image copy recreates it, and for remote images, the ImageStreamTag restore recreates this. Actually, thinking about it, we may still want this resource not restored, since the ImageStreamTag restore will recreate it anyway. @dymurray what do you think we should do in terms of the return value here?
￼

Reply:
Yeah as we discussed in slack I think it makes sense to still not restore this resource since the istag will autorecreate the imagestream. This will help reduce the warnings/errors on Velero's side when running a backup/restore